x/exchange/msg.go

```diff
@@ -0,0 +1,464 @@
+	}
+
+	if err := m.MarketDetails.Validate(); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errors.Join(errs...)
+}
+
+func (m MsgMarketUpdateDetailsRequest) GetSigners() []sdk.AccAddress {
+	addr := sdk.MustAccAddressFromBech32(m.Admin)
+	return []sdk.AccAddress{addr}
+}
+
+func (m MsgMarketUpdateEnabledRequest) ValidateBasic() error {
+	var errs []error
+
+	if _, err := sdk.AccAddressFromBech32(m.Admin); err != nil {
+		errs = append(errs, fmt.Errorf("invalid administrator %q: %w", m.Admin, err))
+	}
+
+	if m.MarketId == 0 {
+		errs = append(errs, fmt.Errorf("invalid market id: cannot be zero"))
+	}
+
+	// Nothing to validate for the AcceptingOrders field.
+
+	return errors.Join(errs...)
+}
+
+func (m MsgMarketUpdateEnabledRequest) GetSigners() []sdk.AccAddress {
+	addr := sdk.MustAccAddressFromBech32(m.Admin)
+	return []sdk.AccAddress{addr}
+}
+
+func (m MsgMarketUpdateUserSettleRequest) ValidateBasic() error {
+	var errs []error
+
+	if _, err := sdk.AccAddressFromBech32(m.Admin); err != nil {
+		errs = append(errs, fmt.Errorf("invalid administrator %q: %w", m.Admin, err))
+	}
+
+	if m.MarketId == 0 {
+		errs = append(errs, fmt.Errorf("invalid market id: cannot be zero"))
+	}
+
+	// Nothing to validate for the AllowUserSettlement field.
+
+	return errors.Join(errs...)
+}
+
+func (m MsgMarketUpdateUserSettleRequest) GetSigners() []sdk.AccAddress {
+	addr := sdk.MustAccAddressFromBech32(m.Admin)
+	return []sdk.AccAddress{addr}
+}
+
+func (m MsgMarketManagePermissionsRequest) ValidateBasic() error {
+	var errs []error
+
+	if _, err := sdk.AccAddressFromBech32(m.Admin); err != nil {
+		errs = append(errs, fmt.Errorf("invalid administrator %q: %w", m.Admin, err))
+	}
+
+	if m.MarketId == 0 {
+		errs = append(errs, fmt.Errorf("invalid market id: cannot be zero"))
+	}
+
+	if m.HasUpdates() {
+		for _, addrStr := range m.RevokeAll {
+			if _, err := sdk.AccAddressFromBech32(addrStr); err != nil {
+				errs = append(errs, fmt.Errorf("invalid revoke-all address %q: %w", addrStr, err))
+			}
+		}
+		if ContainsString(m.RevokeAll, m.Admin) {
+			errs = append(errs, fmt.Errorf("message administrator %s cannot revoke all of their permissions", m.Admin))
+		}
+
+		if err := ValidateAccessGrantsField("to-revoke", m.ToRevoke); err != nil {
+			errs = append(errs, err)
+		}
+
+		toRevokeByAddr := make(map[string]AccessGrant, len(m.ToRevoke))
+		for _, ag := range m.ToRevoke {
+			if ContainsString(m.RevokeAll, ag.Address) {
+				errs = append(errs, fmt.Errorf("address %s appears in both the revoke-all and to-revoke fields", ag.Address))
+			}
+			if ag.Address == m.Admin && ag.Contains(Permission_permissions) {
+				errs = append(errs, fmt.Errorf("message administrator %s cannot revoke their own ability to manage permissions",
+					ag.Address))
+			}
+			toRevokeByAddr[ag.Address] = ag
+		}
+
+		if err := ValidateAccessGrantsField("to-grant", m.ToGrant); err != nil {
+			errs = append(errs, err)
+		}
+
+		for _, ag := range m.ToGrant {
```

Comment:
There is 4 levels of indentation here. Is it possible to restructure this a little bit to make it easier to read? You could add some additional functions or rewrite some of the loop.

```
if !ok {
  continue
}
Happy path...
```

Reply:
I get the desire to not get too tabbed in, but with every best practice, there are exceptions. That short-circuit adds 2 lines and saves a tab on 5 lines. My cutoff is usually around 8-10 tabbed-in lines for doing that. I feel like having only a few lines tabbed in is fine, specially when they're short, and specially when the alternative is either a clunky function extraction or a loop control statement. The one line tabbed in the most is the longest, but it's clear in the first 20 chars that it's just adding an error to the list which is usually all someone needs to know about that line.

Ultimately, I don't think the extra tabs hurt readability here.